### PR TITLE
GET Endpoint for Scorecard Results Retrieval

### PR DESCRIPTION
**Pull request by rohankh532**
Part (2/2) of #33

This is the api endpoint that return the results json for a specified repository and branch.

ref: #32 

Design doc: https://docs.google.com/document/d/18KgkR-5GCkPvcT-wxxSsbsEYQYDbReLM5jBOofwT0l4/edit?usp=sharing&resourcekey=0-tKZEPEifHPHL15-zTACn1Q

### CODE REVIEW

**Comment by rohankh532** [comment]
Ready for review

File: main.go
Comment (azeemshaikh38) [inline main.go:44]: Let's output the usage of the API here. Eg: https://api.github.com/

File: signing/signing.go
Comment (azeemshaikh38) [inline signing/signing.go:26]: Use `filepath.Join(host, orgName, repoName)`

File: signing/signing.go
Comment (azeemshaikh38) [inline signing/signing.go:31]: This bit of code is scary. We should do more validation to check for malicious input. I would suggest starting with `filepath.Clean` to sanitize the path and then use `filepath.Match("*/*/*/results.json")` to ensure that exactly 3 separators exist.

@laurentsimon @asraa in case you have better suggestions.

File: signing/signing_test.go
Comment (azeemshaikh38) [inline signing/signing_test.go:12]: Add tests for testing behavior when URL params contain: "../", "/", "\\/" etc.

**Review by azeemshaikh38 — COMMENTED** [review]
LGTM. Fix the security issues before merging the code in.

File: main.go
Comment (azeemshaikh38) [inline main.go:43]: Use Golang convention for field name: `GetRepoResults`

File: signing/signing.go
Comment (azeemshaikh38) [inline signing/signing.go:31]: No need for these `strings.Replace` call. `filepath.Clean` should already take care of this.

Comment (Rohan6789) [reply signing/signing.go]: Removing these causes CodeQL scanning to fail: https://github.com/ossf/scorecard-webapp/pull/55/checks?check_run_id=5780026266

Comment (rohankh532) [reply signing/signing.go]: Removing these causes CodeQL scanning to fail: https://github.com/ossf/scorecard-webapp/pull/55/checks?check_run_id=5780026266

File: main.go
Comment (laurentsimon) [inline main.go:45]: since the domain name may change, I wonder whether this should all live under a config file and shared across all sub-projects in the webapp repo?

File: main.go
Comment (laurentsimon) [inline main.go:47]: I think you can just use `Marshall()` and there will be no indentation

File: signing/signing.go
Comment (laurentsimon) [inline signing/signing.go:14]: why do we need the SARIF result?

File: signing/signing.go
Comment (laurentsimon) [inline signing/signing.go:35]: code scanning has some useful feedback below. Let's create a function that truncates what we print to log to reduce DoS/log filling; and use this for everything we log

File: signing/signing_test.go
Comment (laurentsimon) [reply signing/signing_test.go:12]: +1, also a test when the bucket does not exist to hit the `http.StatusInternalServerError` path

File: main.go
Comment (rohankh532) [reply main.go:45]: So do we not include the domain name here? Or automatically pull it from somewhere else

Comment (laurentsimon) [reply main.go:45]: yes pull it from a file in the repo. @azeemshaikh38 any suggestion?
Maybe do that in a follow-up PR and create a tracking issue for it.

Comment (azeemshaikh38) [reply main.go:45]: Leave it as `/projects/{host}/{owner}/{repository}` for now. That is generic enough to work for any domain. Leave a `TODO` to make this domain-specific in future.

Comment (laurentsimon) [reply main.go]: my comment was about `api.securityscorecards.dev` though, not the GET request.
If we spin our clusters to answer on all domains, it's fine. But it we stick to a single, having the domain ina config file is better

**Review by azeemshaikh38 — APPROVED** [review]

**Comment by azeemshaikh38** [comment]
LGTM. @laurentsimon for a second review.

File: signing/signing.go
Comment (laurentsimon) [inline signing/signing.go:19]: Using `http.Error` + return seems really prone to mistakes (it's the same way C developers always forget to free arrays). I know that's how the API works, but I think it may be better to do something like:
```
func GetResults(w http.ResponseWriter, r *http.Request) {
results, err := getResults()
if err != nil{
   http.Error(w, err)
   return
}

  w.Header().Set("Content-Type", "application/json")
  w.Write(resultsBytes)
}
```

File: signing/signing_test.go
Comment (laurentsimon) [inline signing/signing_test.go:50]: Do we lack tests for 
1. invalid repoName with valid host and orgName?
2. invalid host but valid orgName and repoName?

File: signing/signing.go
Comment (laurentsimon) [inline signing/signing.go:42]: Would be nice if you could check that you get the proper error in your unit tests. Maybe you could declare the error as:
```
var errorPullingBucket = errors.New("error pulling from GCS bucket")
...
http.Error(w, errorPullingBucket.Error(), http.StatusInternalServerError)
```

for each of these errors.

File: main.go
Comment (rohankh532) [reply main.go:53]: Yeah, fixed.

File: signing/signing.go
Comment (rohankh532) [reply signing/signing.go:35]: CodeQL wanted stronger cleaning: https://github.com/ossf/scorecard-webapp/security/code-scanning/43

But I think filepath.Clean should be enough because it gets rid of all the escaping. Should I just add a nolint to this line?

File: signing/signing_test.go
Comment (rohankh532) [reply signing/signing_test.go:50]: Added them

File: signing/signing.go
Comment (rohankh532) [reply signing/signing.go]: I refactored the code to align with this

File: signing/signing.go
Comment (laurentsimon) [reply signing/signing.go:49]: do you know why there are `\n` and `\r` in the path? Was it because we don't trust the input?

**Review by laurentsimon — APPROVED** [review]
Thanks!

File: signing/signing.go
Comment (laurentsimon) [reply signing/signing.go:55]: ok got it.. sorry I've been slow understanding this :) I suppose it's fine to remove these, just add a comment.
Shall we also truncate the string before logging? Maybe 256 characters max? (We can use the maximum Linux path length, I think). If the string is greater, we can return an error

File: signing/signing.go
Comment (laurentsimon) [inline signing/signing.go:30]: You can make this more generic by simply writing:
```
if err != nil{
  http.Error(w, err.Error(), http.StatusInternalServerError)
  log.Println("err: %v", err)
}
```

`err` already contains the information about the error.
This way you take care of *all* errors with at once

**Review by laurentsimon — APPROVED** [review]
Just 2 comments to address. Thanks!

File: signing/signing.go
Comment (rohankh532) [reply signing/signing.go]: No problem, I added a check for this

**Pull request merged**
